### PR TITLE
[PAYM-166] Create account integration tests with WireMock for Stripe API

### PR DESCRIPTION
**Pull request by mniesyn**
https://lodgify.atlassian.net/browse/PAYM-166

Create base mechanism for integration tests.
Create tests for get accounts and create account scenario.

Rename StripeClient to StripeGatewayClient to avoid name collision.

### CODE REVIEW

File: Lodgify.Payments.Stripe.Server.IntegrationTests/Factories/CustomWebApplicationFactory.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        .WithUsername("postgres")
+        .WithPassword("postgres")
+        .WithPortBinding(8999, true)
+        .Build();
+
+    private readonly WireMockContainer _wireMockContainer = new WireMockContainerBuilder()
+        .Build();
+
+    public IWireMockAdminApi WiremockClient { get; private set; }
+    internal PaymentDbContext DbContext => Services.GetRequiredService<PaymentDbContext>();
```
Comment (marcinmilik) [inline Lodgify.Payments.Stripe.Server.IntegrationTests/Factories/CustomWebApplicationFactory.cs:39]: You are creating this DbContext without creating a new scope. Is it correct?

Meaning shouldn't we have a method that returns disposable scope and from this scope PaymentDbContext is reasolved, all operations on it are done and after that the scope is disposed?

Otherwise we do not have guaranty that we use dbcontext from isolated new scope.

File: Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+using WireMock.Admin.Mappings;
+using WireMock.Client;
+using Xunit;
+
+namespace Lodgify.Payments.Stripe.Server.IntegrationTests.Shared;
+
+[CollectionDefinition(nameof(PaymentsCollection))]
+public class PaymentsCollection : ICollectionFixture<CustomWebApplicationFactory>;
+
+[Collection(nameof(PaymentsCollection))]
```
Comment (marcinmilik) [inline Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs:18]: Aren't tests from the same collection executed sequentially?
Wouldn't we like to be able to run tests in parallel?
Maybe it is better to define collection and also fixture on the main class level:
```
[Collection(nameof(PaymentsCollection))]
public class AccountControllerTests : BaseIntegrationTest

```

File: Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+public abstract class BaseIntegrationTest
```
Comment (marcinmilik) [inline Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs:19]: Would it be a base class for all tests? Or only these that require Stripe API exposed behind WireMock?
Because of that how about:
 ```suggestion
public abstract class WireMockStripeIntegrationTest
```

File: Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+using Lodgify.Payments.Stripe.Server.IntegrationTests.Helpers;
+using Microsoft.EntityFrameworkCore;
+using WireMock.Admin.Mappings;
+using WireMock.Client;
+using Xunit;
+
+namespace Lodgify.Payments.Stripe.Server.IntegrationTests.Shared;
+
+[CollectionDefinition(nameof(PaymentsCollection))]
+public class PaymentsCollection : ICollectionFixture<CustomWebApplicationFactory>;
```
Comment (marcinmilik) [inline Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs:16]: It's related to the previous comments.
What test collections/fixtures we'd like to have? One for all or multiply more granular?
If the second option then how about:
```suggestion
public class AccountManagmentCollection : ICollectionFixture<CustomWebApplicationFactory>;
```

File: Lodgify.Payments.Stripe.Server/Program.cs [hunk 5]
```diff
@@ -30,4 +30,9 @@
 finally
 {
     LogManager.Shutdown();
-}
\ No newline at end of file
+}
+
+namespace Lodgify.Payments.Stripe.Server
+{
+    public partial class Program { }
```
[the remark this answers is not in the document]
Comment (mniesyn) [reply Lodgify.Payments.Stripe.Server/Program.cs:37]: this partial is necessary only for integration tests and it is follow with Microsoft recommendations https://learn.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-8.0

File: Lodgify.Payments.Stripe.Server.IntegrationTests/Factories/CustomWebApplicationFactory.cs [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        await migrator!.MigrateAsync(pendingMigrations.Last(), CancellationToken.None);
+    }
+
+    private async Task SeedDatabaseAsync()
+    {
+        using var scope = Services.CreateScope();
+        var dbContext = scope.ServiceProvider.GetRequiredService<PaymentDbContext>();
+
+        var transaction = await dbContext.Database.BeginTransactionAsync(CancellationToken.None);
+        await transaction.CommitAsync(CancellationToken.None);
```
[the remark this answers is not in the document]
Comment (mniesyn) [reply Lodgify.Payments.Stripe.Server.IntegrationTests/Factories/CustomWebApplicationFactory.cs:122]: No no, for a moment it is a "placeholder". However I was added it just to be clear for anybody where to populate db by some test data if necessary. I recommend to leave it as it is.

File: Lodgify.Payments.Stripe.Server.IntegrationTests/Factories/CustomWebApplicationFactory.cs [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        integrationConfig["ConnectionStrings:Postgres"] = _postgresContainer.GetConnectionString();
+
+        var wireMockUrl = _wireMockContainer.GetPublicUrl();
+        integrationConfig["Identity:BaseUrl"] = wireMockUrl;
+        integrationConfig["StripeSettings:ApiBase"] = wireMockUrl;
+
+        return integrationConfig;
+    }
+
+    private async Task MigrateDatabaseAsync(Type dbContextType)
```
[the remark this answers is not in the document]
Comment (mniesyn) [reply Lodgify.Payments.Stripe.Server.IntegrationTests/Factories/CustomWebApplicationFactory.cs:102]: It could be :) this is only kind of helper method for tests but why not.

File: Lodgify.Payments.Stripe.Server.IntegrationTests/Factories/CustomWebApplicationFactory.cs [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        using var scope = Services.CreateScope();
+        var dbContext = (scope.ServiceProvider.GetRequiredService(dbContextType) as DbContext)!;
```
[the remark this answers is not in the document]
Comment (mniesyn) [reply Lodgify.Payments.Stripe.Server.IntegrationTests/Factories/CustomWebApplicationFactory.cs:105]: That will change nothing. The compiler will still complain about possible null reference because GetRequiredService method return nullable object.

File: Lodgify.Payments.Stripe.Server.IntegrationTests/Factories/CustomWebApplicationFactory.cs [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+        var pendingMigrations = await dbContext.Database.GetPendingMigrationsAsync(CancellationToken.None);
+
+        var migrator = dbContext
+            .GetInfrastructure()
+            .GetService(typeof(IMigrator)) as IMigrator;
```
[the remark this answers is not in the document]
Comment (mniesyn) [reply Lodgify.Payments.Stripe.Server.IntegrationTests/Factories/CustomWebApplicationFactory.cs:111]: The same here if we really want to stop compiler complain then it should looks like:
  var migrator = (IMigrator)(dbContext
            .GetInfrastructure()
            .GetService(typeof(IMigrator)))**!**;

File: Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs (hunk 3 above)
Comment (mniesyn) [reply Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs:19]: My idea was make it common for all because it contains helper method like Insert to db, like deserialize response which is not realated with Stripe or WireMock. Then current solution still make sense for you?

File: Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs (hunk 4 above)
Comment (mniesyn) [reply Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs:16]: Hmmm not really get the point here... to be sure we are on the same page the FixtureCollection simply share context among the tests.  It makes sense to have multiple collections if you need to share different contexts or setups among different groups of tests. Each collection can have its own setup and teardown logic, which can be useful for isolating tests that require different configurations or dependencies.  For example, you might have one collection for tests that require a real Stripe setup and another for tests that require another configuration. This helps in organizing tests and ensuring that the setup and teardown logic is only applied to the relevant tests, improving test isolation and reducing potential side effects.
So to sum up, personally I don`t think so we want be so granullar and create separate collection per stripe operation, my idea was to have one fixture collection for mocks and one for real connection with stripe. In this second option we can then do stuff like:

```
public class StripeApiFixture : WebApplicationFactory<Program>, IDisposable
{
    protected override void ConfigureWebHost(IWebHostBuilder builder)
    {
        builder.ConfigureServices(services =>
        {
            // Configure test services here
        });
    }

    public void Dispose()
    {
        // Clean up stripe resources
    }
}
```

What you think?

File: Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs (hunk 2 above)
Comment (mniesyn) [reply Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs:18]: Base on my knowlege:
- tests in the **same** collection are run sequentially 
- tests in another collections are run paraller

Also because of AccountControllerTests already inherit from BaseIntegrationTest it is not necessary to additionally decorate class.

So to summarize, this eventually additionall class decoration doesn`t have impact on this if test are run paraller on not, the assign to appropriate collection has matter. 
The question is do we want create separate collection just to make group of tests run in paraller? Personally for me now it sounds like overengeenering, especially we have one/two tests in collection.

File: Lodgify.Payments.Stripe.Server.IntegrationTests/Factories/CustomWebApplicationFactory.cs (hunk 1 above)
Comment (mniesyn) [reply Lodgify.Payments.Stripe.Server.IntegrationTests/Factories/CustomWebApplicationFactory.cs:39]: Thats good point!
I was already address it

File: Lodgify.Payments.Stripe.Server.IntegrationTests/Factories/CustomWebApplicationFactory.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (marcinmilik) [reply Lodgify.Payments.Stripe.Server.IntegrationTests/Factories/CustomWebApplicationFactory.cs:123]: got it. how about adding a comment that explains that it is place holder and generic seed data for all tests that use this factory has to be put there?

File: Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs (hunk 2 above)
Comment (marcinmilik) [reply Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs:18]: My point is that currently the code is prepared in the way to run all tests that inherits from BaseIntegrationTest in the scope of the same collection. So all future tests will be sequentially executed.

> The question is do we want create separate collection just to make group of tests run in paraller? Personally for me now it sounds like overengeenering, especially we have one/two tests in collection.

My understanding from our groomings was that from the beginning we want to have new tests run in parallel. I was thinking about this code as a base foundation for doing it.

We could start from sequential with the idea to refactor later to parallel once we have significant number of tests.
There is a risk that we may spot issues with running tests at the same time and this refactoring may be bigger.

> Also because of AccountControllerTests already inherit from BaseIntegrationTest it is not necessary to additionally decorate class.

I think I was not clear. I was meaning decorate AccountControllerTests instead of BaseIntegrationTest so person who is writing a new test will need to decide if he/she wants to run them in the same collection. So even if now we will have tests in the same collection soon we could consider adding other collections.

What do you think?

File: Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs (hunk 3 above)
Comment (marcinmilik) [reply Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs:19]: Hmm you coupled `CustomWebApplicationFactory` with `BaseIntegrationTest` and `CustomWebApplicationFactory` runs the application with wiremock setup as a stripe API. So how do you plan to introduce tests that call real Stripe API?

File: Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs (hunk 4 above)
Comment (marcinmilik) [reply Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs:16]: I see. I think it is related to the previous topic. 
How we want to design our tests in terms of collections/parallel execution?

My way of thinking was to run tests related to us calling Stripe for update/create Account with one collection (AccountApi / AccountManagment), tests related to Stripe calling us to update Account through webhook with other collection etc.

I am afraid that we may develop the solution that forces us to have single base class with single fixture and all tests will be executed in the same collection because it will be too much work to introduce new collection.

**Comment by marcinmilik** [comment]
I see that tests do not want to run in CI https://jenkins.lodgify.net/blue/organizations/jenkins/lodgify-payments-stripe/detail/PR-19/2/pipeline please check if all needed test related nugets are installed in new integration csproj

File: Lodgify.Payments.Stripe.Server.IntegrationTests/Factories/CustomWebApplicationFactory.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mniesyn) [reply Lodgify.Payments.Stripe.Server.IntegrationTests/Factories/CustomWebApplicationFactory.cs:123]: Lets leave it because in next task it will be used so no any additionall comments will be needed

File: Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs (hunk 2 above)
Comment (mniesyn) [reply Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs:18]: Concept is ok, however lets leave it in current scope and in next task (with real stripe integration) there will be good reason to split it

File: Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs (hunk 4 above)
Comment (mniesyn) [reply Lodgify.Payments.Stripe.Server.IntegrationTests/Shared/BaseIntegrationTest.cs:16]: No, no it will be splitted on few collection, let`s move it to the scope of next task (real stripe integration) just to make real example of it

**Review by marcinmilik — APPROVED** [review]
